sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -129,17 +131,41 @@ abstract class BinaryArithmetic extends BinaryOperator with NullIntolerant {
   def calendarIntervalMethod: String =
     sys.error("BinaryArithmetics must override either calendarIntervalMethod or genCode")
 
+  def checkOverflowCode(result: String, op1: String, op2: String): String =
+    sys.error("BinaryArithmetics must override either checkOverflowCode or genCode")
+
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = dataType match {
     case _: DecimalType =>
       defineCodeGen(ctx, ev, (eval1, eval2) => s"$eval1.$decimalMethod($eval2)")
     case CalendarIntervalType =>
       defineCodeGen(ctx, ev, (eval1, eval2) => s"$eval1.$calendarIntervalMethod($eval2)")
+    // In the following cases, overflow can happen, so we need to check the result is valid.
+    // Otherwise we throw an ArithmeticException
     // byte and short are casted into int when add, minus, times or divide
     case ByteType | ShortType =>
-      defineCodeGen(ctx, ev,
-        (eval1, eval2) => s"(${CodeGenerator.javaType(dataType)})($eval1 $symbol $eval2)")
+      nullSafeCodeGen(ctx, ev, (eval1, eval2) => {
+        val overflowCheck = if (checkOverflow) {
+          checkOverflowCode(ev.value, eval1, eval2)
+        } else {
+          ""
+        }
+        s"""
+           |${ev.value} = (${CodeGenerator.javaType(dataType)})($eval1 $symbol $eval2);
+           |$overflowCheck
+         """.stripMargin
+      })
     case _ =>
-      defineCodeGen(ctx, ev, (eval1, eval2) => s"$eval1 $symbol $eval2")
+      nullSafeCodeGen(ctx, ev, (eval1, eval2) => {
+        val overflowCheck = if (checkOverflow) {
+          checkOverflowCode(ev.value, eval1, eval2)
+        } else {
+          ""
+        }
+        s"""
+           |${ev.value} = $eval1 $symbol $eval2;
```

Comment:
Is it more efficient to call `Math.addExact` instead of adding an extra check?

Reply:
yes, this is good point actually. As I mentioned in https://github.com/apache/spark/pull/21599#issuecomment-505924180, the issue with this is that `addExact` exists only for few datatypes. So we should specialize many cases in the code here. So here we have a tradeoff between performances (in the check overflow case only) and code maintainability. What is your opinion on this?